### DISCUSSION TITLE
Vitest Browser Mode

### DISCUSSION THREAD

**Discussion by sheremet-va** [vitest-dev/vitest#5828] [Announcements]
It's been some time since we gave any update on the Browser Mode, and Vitest 2.0 seems like a good time to explain how we feel about it. And we have a lot of plans!

The initial purpose of Vitest was to allow running tests without a complex Jest setup but with the battle-tested Jest API everyone is already familiar with. This also meant faking the browser environment (with [`jsdom`](https://github.com/jsdom/jsdom) or [`happy-dom`](https://github.com/capricorn86/happy-dom)). With time, as the JavaScript ecosystem evolved, we started to hit roadblocks with this approach (like this issue, for example, https://github.com/vitest-dev/vitest/issues/4043). Switching to Jest's VM approach allowed us to bypass some of them, but introduced many more along the way (https://github.com/vitest-dev/vitest/pull/3203).

Using fake browser environments means that developers introduce an environment that doesn't exist anywhere except tests which can lead to testing something that cannot happen when an actual person uses your Web Application. Fake environments also have their own limitations like not providing common Web APIs (`fetch`, `canvas`, `PerformanceObserver`, `HTMLDialogElement`, Pointer Lock, and so on) - this means that to test a single component you have to resort to E2E testing which throws a lot of APIs at you and can require a complex setup. There are already several solutions to that by different testing frameworks:

- [Playwright Components](https://playwright.dev/docs/test-components)
- [Cypress Component Testing](https://docs.cypress.io/guides/component-testing/overview)
- [Webdriverio Component Testing](https://webdriver.io/docs/component-testing/)

[Browser Mode](https://main.vitest.dev/guide/browser) is our take on component testing. 

## Future Plans

Browser mode aims to replace `jsdom` and `happy-dom` environment with an easy migration path. Right now you are able to opt-in to using browser mode by creating a separate [workspace project](https://vitest.dev/guide/workspace):

```ts
// vitest.workspace.ts
import { defineWorkspace } from 'vitest/config'

export default defineWorkspace([
  './vitest.config.ts', // the default config for Node.js tests
  {
    test: {
      include: ['**/*.browser.test.ts'],
      browser: {
        enabled: true,
        name: 'chrome',
      },
    },
  },
])
```

When you switch to the browser mode, all your tests should work as before unless you use Node.js APIs directly. This means that tests will be isolated by file, not by the `test` function (similar to Web Test Runner or Cypress). Most of the Vitest APIs including [`vi.mock`](https://github.com/vitest-dev/vitest/pull/5765), `vi.dynamicImportSettled` should just work. Some of them we are unable to implement due to ESM limitations like `resetModules` - these will throw an error when running in the browser.

The Vitest team is not interested in exploring the E2E testing (Vitest already provides integration for `webdriverio` and `playwright` that you can use alongside Vitest Browser Mode). In E2E tests you are typically testing a site that is deployed and doesn't need any changes. In Vitest Browser Mode you provide what is mounted on the DOM and mock the modules you want. Changing different states of the DOM becomes easier. This also means that we will not provide higher-level APIs like network interception. (On that note, our current recommendation is to use [MSW](https://mswjs.io/) for both browser and Node.js tests - this will be reflected in the browser guide in our documentation).

It's hard to start using the browser mode in CI at the moment because you need to know what provider to choose and how to configure Vitest. To address this, we will improve the documentation and also [provide a command](https://github.com/vitest-dev/vitest/issues/5793) to install all necessary packages and update the configuration to use the browser mode.

## Additional API

Vitest will provide the context object via `@vitest/browser/context` import that exposes utilities (like file reading/writing) and allows you to communicate with the Node.js server via the [Commands API](https://github.com/vitest-dev/vitest/pull/5097). The context will expose the `page` object with interactive APIs that should replace the `userEvent` from `@testing-library/user-event`. The current idea is to replace the `@testing-library/user-event` import with a [module that exposes the way to interact with the browser](https://github.com/vitest-dev/vitest/issues/5770) through Chrome DevTools or Webdriver protocols instead of mocking events that might not follow the actual browser specification:

```ts
// if imported from `@testing-library/user-event'
// we intercept it to use CDP under the hood instead
import { userEvent } from '@vitest/browser/context'
import { findByText } from '@testing-library/dom'

it('clicks correctly', async () => {
  await userEvent.click(await findByText('Click'))
})
```

Vitest will also [provide a way](https://github.com/vitest-dev/vitest/issues/5812) to hook into CDP directly to make developing libraries easier.

By default, Vitest provides [playwright](https://playwright.dev/) and [webdriverio](https://webdriver.io/) APIs wrapped in a common API out of the box (depending on which one you use) unless you use `preview` provider (previously named `none`).

```ts
import { page, commands, server } from '@vitest/browser/context'

page.config // ResolvedConfig
page.browser // chrome
server.platform // node
await commands.readFile('./relative-to-test-file.js')
```

The `preview` provider (called `none` in Vitest 1.6) allows running tests in any browser of your choice by simply opening a link. It exposes a subset of APIs available in `playwright` and `webdriverio`, and uses the actual `userEvent` library to _simulate_ events without relying on DevTool or Webdriver protocols. This provider is useful for a quick inspection of your code and tests and it doesn't require installing additional libraries and browsers, it is also the only provider that works in [Stackblitz](https://stackblitz.com/). However, be aware that if you don't use it to run tests in Stackblitz, your browser extensions/cookies/localStorage will affect your tests and possibly the dev environment making it hard to debug potential issues with tests. We recommend installing `playwright` or `webdriverio` to run your tests locally.

Retryability is important when testing in the browser. Vitest 2.0 provides new `expect.poll` API to retry your assertions manually:

```ts
import { screen } from '@testing-library/dom'
import '@testing-library/jest-dom'

const input = await screen.findByRole('input', { name: 'Username' })
await expect.poll(() => input).toHaveFocus()
```

Currently, there are no plans to implement our own built-in retryability+query API, but this can change based on community feedback.

Viewport can be configured globally with the `browser.viewport` option, but you can also change it during your test by clicking on one of the presets in the UI or by using `page.viewport` command:

```ts
import { page } from '@vitest/browser/context'
import { it } from 'vitest'

it('viewport changes', async () => {
  await page.viewport(300, 500)
})
```

## Providers

Vitest requires `playwright` or `webdriverio` provider to run tests without simulating any events. Providers are used to install browsers, connect to CDP, and expose the testing browser environment.

Vitest doesn't use test runners from these providers and doesn't expose any of their APIs directly. Built-in `userEvent` uses `page` and `$` APIs under the hood depending on the provider. You can write [Custom Commands](https://main.vitest.dev/guide/browser.html#commands) to get access to providers APIs on the server. Beware that you might need to use `page.frameLocator` or `browser.switchToFrame` APIs to get access to HTML elements. This is considered an advanced use case, and we do not encourage overusing it.

## Test Isolation

Vitest opens a single page during the test run. Tests are isolated using iframes inside of that page. During development, you can see the iframe inside the UI (configurable by `test.browser.ui`):

<img width="1392" alt="Running tresjs example in Vitest BrowserMode" src="https://github.com/vitest-dev/vitest/assets/16173870/3b6bc721-b194-416d-a685-cf7282ee802a">

_Running [tresjs](https://tresjs.org/) example in Vitest Browser Mode_

The `--no-isolate` flag also allows running all browser tests inside the same iframe in cases where you do not depend on the global state and do a proper cleanup. This can make your test run several times faster. Module mocking works even when isolation is disabled.

## Debugging

Debugging using the IDE is not supported at the moment, but it's on the roadmap. In the meantime, you can use the `debugger` statement in your tests to automatically stop execution and open devtools:

```ts
import { it } from 'vitest'

it('stops execution', () => {
  debugger;
})
```

Unfortunately, you might need to rerun the test several times to hit the debugger statement if they are too fast. Better debugging experience is one of the highest priorities for the next feature release.

## Node.js Polyfills

Vitest doesn't polyfill any Node.js APIs because it would create a situation where your code will work in tests, but won't work in dev or production. If you rely on some Node.js APIs in your setup files, you can add [`vite-plugin-node-polyfills`](https://github.com/davidmyersdev/vite-plugin-node-polyfills) to your plugins list if `process.env.VITEST` is set and manually add them to `include` option:

```ts
import { nodePolyfills } from 'vite-plugin-node-polyfills'
import { defineConfig } from 'vitest/config'

export default defineConfig({
  plugins: [
    process.env.VITEST ? nodePolyfills({
      include: ['path']
    }) : null,
  ],
})
```

## Current State

Vitest 1.6 implements browser tests using iframes. During the test run, you will see the [Vitest UI](https://vitest.dev/guide/ui.html), changing the file will reload the browser page.

[Vitest 2.0](https://github.com/vitest-dev/vitest/discussions/5210) will ship with [Commands API](https://github.com/vitest-dev/vitest/pull/5097) and [module mocking](https://github.com/vitest-dev/vitest/pull/5765) supported. You can already test changes using the [beta branch](https://www.npmjs.com/package/vitest/v/2.0.0-beta.7).

This release also implements better watch mode support - you can change files directly inside the UI and the tests will be reloaded without a page refresh like it was in Vitest 1.6. Changing the file outside the UI will also not reload the page.

The [interactivity API](https://github.com/vitest-dev/vitest/issues/5770) is planned to be released after Vitest 2.0.

Special thanks to wonderful @Aslemammad and @userquin who worked on the initial implementation 🎉

**Comment by bahmutov** [comment]
Interesting, reading this with excitement

**Comment by NoriSte** [comment]
Amazing, can't wait for all the next evolutions!!

**Comment by shreyT19** [comment]
Incredible!

**Comment by ishpreet95** [comment]
shut up and take my money

**Comment by favna** [comment]
Maybe mention https://testcafe.io in your list of E2E frameworks?

**Reply by sheremet-va** [reply]
We don't have the list of E2E frameworks, only the list of component testing solutions. Does testcase support component testing?

**Reply by favna** [reply]
Ah no my bad, I misread then. Testcafe is purely E2E.

**Comment by JessicaSachs** [comment]
Pretty stoked to see this work. I really appreciate being consulted about the future Vitest Browser Testing arch (compared to Cypress Component Testing's arch). I'm confident that this architecture will meet my needs for Vite-specific apps and I'll undoubtedly phase out teaching Cypress Component Testing once this lands.

**Comment by cqh963852** [comment]
Greeting, I am trying to use the `2.0.0-beta.9` version.

rendering a canvas element by "@testing-library/react" 's render.

but the `toMatchInlineSnapshot` did not work as expected.

Is there anything I need to pay attention to here?

```
import { Stage } from "@idealjs/pixi-react";
import { render } from "@testing-library/react";
import { StrictMode, Suspense } from "react";
import { describe, expect, it } from "vitest";

import ECSEntity from "../../../../../ecs-react/src/ECSEntity";
import ECSWorld from "../../../../../ecs-react/src/ECSWorld";
import RenderCellComponent from "../../components/RenderCellComponent";
import RenderSystem from ".";

describe("RenderSystem", () => {
  it("RenderCellComponent", async () => {
    const { container } = render(
      <StrictMode>
        <Suspense>
          <Stage>
            <ECSWorld>
              <RenderSystem />
              <ECSEntity>
                <RenderCellComponent ...
                />
              </ECSEntity>
            </ECSWorld>
          </Stage>
        </Suspense>
      </StrictMode>,
    );
    expect(container).toMatchInlineSnapshot();
    console.log(container);
  });
});
```

![image](https://github.com/vitest-dev/vitest/assets/17702287/ced219d0-e009-4ead-ae12-ee2af2187401)

![image](https://github.com/vitest-dev/vitest/assets/17702287/effd9c7e-8cda-4824-91a0-9d80b4a283a6)

**Reply by sheremet-va** [reply]
`toMatchInlineSnapshot` is bugged in beta at the moment. You can create an issue with a small reproduction.

**Comment by abenhamdine** [comment]
Will it be possible to set a custom render function to mount component from another framework than React ?
in other words, is Vitest component testing framework-agnostic ?
We have a app running on a legacy js framework  and we would want to do component testing with Vitest.

**Reply by sheremet-va** [reply]
Vitest doesn't provide any method to mount components. You are encouraged to use `@testing-library/*` package depending on your framework or render it yourself however you want.

**Reply by abenhamdine** [reply]
thanks for your answer, indeed @testing-library/dom will be enough for our case.

**Comment by cqh963852** [comment]
~~I encountered some problems when customizing commands.~~

~~How should I set the provider type so that provider.page can be used without type error?~~

---

The above question is not important, so I modified it

**Reply by cqh963852** [reply]
~~This problem was solved after putting `vitest.config.ts` into the `include ` of `tsconfig.json`~~

**Reply by cqh963852** [reply]
~~I found that current page type of playwright's provider is not fully Page.~~

~~So that I can't access the screenshot function.~~

~~I have create a [PR](https://github.com/vitest-dev/vitest/pull/5874), PTAL~~

**Reply by cqh963852** [reply]
@sheremet-va Thanks for review. I saw the comment in the PR.  

Any plans to expose the `provider.page` ? Or are there plans for support for screenshots?

Canvas related tests usually cannot be performed in jsdom and can only be postponed to the e2e testing stage. 
If vitest2.0 browser mode can support taking relevant screenshots in unit tests, it can solve many problems.

**Reply by sheremet-va** [reply]
Yes, `page` and `tester` (iframe) will be exposed on the context directly

**Reply by sheremet-va** [reply]
The beta 10 now implements parallelization (only for playwright for now).

`page`/`tester`/`body` are exposed in commands context for the playwright provider: https://main.vitest.dev/guide/browser.html#custom-playwright-commands

This release also adds `page.screenshot` command directly to the context object:

```ts
import { page } from '@vitest/browser/context'

it('some test', async () => {
  const screenshotPath = await page.screenshot()
  await page.screenshot({ element: document.querySelector('.element') })
})
```

Playwright provider doesn't have the `page` on its public provider API anymore because tests are running in parallel and there are several pages.

**Reply by tylerbmyers** [reply]
First off, thanks for working on all these improvements to Browser Testing - it is extremely exciting to get all of the incredible features of Vitest with real browser testing! Would the only way to use Playwright's locators be in a custom command? The reason I ask is that I work on an application that uses mostly web components from a design system and testing them has always been a pain except in Playwright. Playwright's locators work by **default** with the [Shadow DOM](https://playwright.dev/docs/locators#locate-in-shadow-dom) and being able to use these rather than `testing-library` which does not support the shadow dom and does not seem to have any plans to support would be another massive step forward.

**Reply by sheremet-va** [reply]
For now, yes. I have some ideas to introduce a way to have Playwright locators built-in, but this is probably for Vitest 2.1 or later.

**Reply by oscarmarina** [reply]
Hi @sheremet-va, as @tylerbmyers  mentioned, thanks for your dedication to improving Browser Testing. It's exciting to see all the advancements being made in this area.

I've been actively exploring Vitest in combination with [Lit -Web Components](https://lit.dev/)  and wanted to share my experience. I've set up a repository where I compare the testing capabilities of Lit using Vitest Browser and Playwright.

 - Repo: [Lit Vitest Testing Comparison](https://github.com/oscarmarina/lit-vitest-testing-comparison/) 
 - StackBlitz: [Lit Vitest Testing Comparison](https://stackblitz.com/github/oscarmarina/lit-vitest-testing-comparison?file=test%2Fcounter-element.test.js)

 > [TS - branch](https://github.com/oscarmarina/lit-vitest-testing-comparison/tree/feature/typescript)

I'm truly impressed with the work being done in Vitest. As you can see in the repo, I've seamlessly integrated utilities from Web Test Runner like AXE and Vitest's new context module, all while leveraging Playwright without any issues.

**Comment by rynoV** [comment]
Thanks for the detailed update!

Do you have any idea how easy it would be to support electron as an option for browser mode? 

I was hoping after seeing the [electron config](https://webdriver.io/docs/desktop-testing/electron/configuration) for webdriverio that it wouldn't be too difficult, but I'm not very familiar yet with how `webdriverio` and browser mode work under the hood, so I'm hoping these are not fundamentally incompatible

**Reply by sheremet-va** [reply]
I have not tested the electron yet, but I know that playwright has experimental support for it, we can try to integrate it if there is demand (I recommend opening an issue).

**Reply by rynoV** [reply]
Issue here https://github.com/vitest-dev/vitest/issues/5883

**Comment by PindaPixel** [comment]
Any way to let the browser runner serve static files (in my case MSW's serviceWorker.js)?

**Reply by sheremet-va** [reply]
Browser Mode is just a Vite server, it serves file from `/public` folder (or configured [`publicDir`](https://vitejs.dev/config/shared-options.html#publicdir)) as static files.

**Reply by edwincarlflores** [reply]
@PindaPixel I cannot seem to make MSW browser worker work with browser mode. Do you have an example repo where MSW is working? Currently using MSW + React testing library

**Reply by ysgk** [reply]
In my case, I had to call `worker.start()` in the test file instead of in the setup file.

**Reply by edwincarlflores** [reply]
@ysgk Did you just added the mockServiceWorker.js file in the public directory for vitest server to pick it up?

**Reply by ysgk** [reply]
Yes. Are you unable to access /mockServiceWorker.js when running the vitest browser?
https://mswjs.io/docs/integrations/browser#generating-the-worker-script

**Comment by cqh963852** [comment]
Hi, I found that `in source` test, not work with browser mode.

Are there any restrictions ?

**Comment by cqh963852** [comment]
While using `expect(await page.screenshot()).toMatchSnapshot();`, vitest will create a snap file has absolute path.

I tried using addSnapshotSerializer in the setup file to handle this path. but I can't access `__dirname`.

Is there a better way to handle paths?

**Reply by cqh963852** [reply]
I found that the browser API is available.
```
const __filename = new URL("", import.meta.url).pathname;
// Will contain trailing slash
const __dirname = new URL(".", import.meta.url).pathname;
```

**Reply by sheremet-va** [reply]
`expect(await page.screenshot()).toMatchSnapshot();` doesn't make any sense. The file path depends on the test name, not the content of the screenshot, so it will never fail.

**Reply by sheremet-va** [reply]
At the moment, Vitest doesn't provide any way to have visual regression testing out of the box, the `screenshot` method just makes a screenshot, there is no mechanism to compare it.

**Reply by cqh963852** [reply]
> The file path depends on the test name, not the content of the screenshot, so it will never fail.

Thanks for your reply, I'm curious about this question

If I want to customize SnapshotSerializer to get the ability to compare pictures. 

How can I get the new picture data and the old picture data?

---

The old pictures are a little troublesome. I think I can write the pictures to the public folder. In this way, when the test program is running, I can get the corresponding pictures.

Can I get the data of the new picture through addSnapshotSerializer?

**Reply by sheremet-va** [reply]
You cannot create a custom snapshot serializer to compare images because it needs to be async, and serializers are always sync.

I don't have any advice on how to implement visual testing.

**Reply by cqh963852** [reply]
I found that playwright's `screenshot` and webdriver's `saveScreenshot` both return a `Promise<Buffer>`

Can vitest's `page.screenshot` return a `Promise<Buffer>` instead `string` path. Or is there any function that can support this ?

So that I can compare It like

```
expect(await page.bufferScreenshot()).toMatchSnapshot({ path:"xxx" });
```

**Reply by sheremet-va** [reply]
Comparing buffers also doesn't make any sense because the same element can produce different buffers depending on the machine, time of day, browser, phase of the moon, the mood of your uncle, American president, or pixel density. To compare images, you need to use external libraries like [pixelmatch](https://github.com/mapbox/pixelmatch).

I will add a [`base64` property](https://github.com/vitest-dev/vitest/pull/5993) because those tools can use it, but I do not recommend comparing base64 strings.

**Comment by njzydark** [comment]
In browser mode, it seems that the iframe is destroyed every time after executing the test. Can we keep the iframe so that I can inspect the component's UI

**Reply by sheremet-va** [reply]
The last iframe is not destroyed, so you can always inspect the latest test. What can be destroyed is the DOM if you use testing-library because it automatically cleans up state - in that case I would recommend disabling auto cleanup and do the cleanup yourself in `beforeEach` hook instead - so the state is kept when you are done

**Comment by sheremet-va** [comment]
I am going to lock the thread now. If you have any issues, bugs, or suggestions, please open a [new issue](https://github.com/vitest-dev/vitest/issues/new/choose) or [start a discussion](https://github.com/vitest-dev/vitest/discussions/new/choose).

I will keep updating this thread whenever we want to release a significant change, so you can keep it in your subscriptions.

**Comment by sheremet-va** [comment]
Vitest 2.1 will have some major improvements to the Browser Mode.

We decided to have a built-in querying mechanism based on Playwright's [locators](https://main.vitest.dev/guide/browser/locators.html) - this API is available in all providers (`preview` and `webdriverio` also).

```ts
await expect.element(
  page.getByRole('heading', { name: 'Sign up' })
).toBeVisible()

await page.getByRole('textbox', { name: 'Login' }).fill('admin')
```

- Locators support Shadow DOM
- Locators work well with a built-in retry-ability mechanism - you can use `await expect.element(locator)` to query the DOM until the element is found
- Documentation was updated to reflect [the new API](https://main.vitest.dev/guide/browser/#examples). We no longer recommend using `testing-library` for every framework - instead, we opted-in to create wrappers around rendering mechanism for several popular frameworks ([vitest-browser-vue](https://github.com/vitest-dev/vitest-browser-vue), [vitest-browser-svelte ](https://github.com/vitest-dev/vitest-browser-svelte) and [vitest-browser-react](https://github.com/vitest-dev/vitest-browser-react)). Feel free to expand this list with your own wrappers!
- [The documentation](https://main.vitest.dev/guide/browser/) in general was rewritten to reduce confusion and provide more information for core functionality

No news on the release date yet, but you can already use these APIs in [the beta](https://github.com/vitest-dev/vitest/releases). Looking forward to your feedback!

**Comment by sheremet-va** [comment]
Vitest 2.1.4 introduces a new option called [`testerHtmlPath`](https://vitest.dev/config/#browser-testerhtmlpath). It is a path to the custom HTML entry point. It can be relative to the root of the project. 

```ts
// vitest.config.ts
import { defineConfig } from 'vitest/config'

export default defineConfig({
  test: {
    browser: {
      testerHtmlPath: './my-custom-document.html',
    },
  },
})
```

```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <!-- custom meta tags -->
  </head>
</html>
<body>
  <!-- prepared HTML structure -->
  <!-- vitest will inject your test file here -->
</body>
```

Vitest will use the HTML content of this file in the tester iframe. It means that all your tests have access to it. If you want to use several HTML files, you can define a custom [workspace](https://vitest.dev/guide/workspace.html) with different project configs:

```ts
// vitest.workspace.ts
import { defineWorkspace } from 'vitest/config'

export default defineWorkspace([
  {
    test: {
      browser: {
        testerHtmlPath: './my-custom-document.html',
      },
    },
  },
  {
    test: {
      include: ['*.special.test.ts'],
      browser: {
        testerHtmlPath: './my-special-document.html',
      },
    },
  },
])
```

The HTML content is processed with the plugins' [`transformIndexHtml`](https://vite.dev/guide/api-plugin.html#transformindexhtml) hooks. This means that you can also inject HTML tags via the plugin API instead of creating a separate file:

```ts
// vitest.config.ts
import { defineConfig } from 'vitest/config'

export default defineConfig({
  plugins: [
    {
      transformIndexHtml() {
        return [
          {
            tag: 'script',
            children: 'windows.test = true',
          },
        ]
      }
    },
  ],
})
```
